crates/pipeline-manager/src/compiler/rust_compiler.rs

```diff
@@ -347,6 +428,186 @@ impl From<UtilError> for RustCompilationError {
+            upload_binary_to_endpoint_with_retries(
+                common_config,
+                delivery_mode,
+                source_file_path,
+                metadata,
+            )
+            .await
+        }
+    }
+}
+
+/// Uploads the compiled binary to an HTTP endpoint with retry logic.
+async fn upload_binary_to_endpoint_with_retries(
+    common_config: &CommonConfig,
+    delivery_mode: &BinaryDeliveryMode,
+    binary_path: &Path,
+    metadata: &BinaryMetadata,
+) -> Result<String, RustCompilationError> {
+    let mut attempts = 0;
+
+    let (endpoint, timeout_secs, max_retries, mut retry_delay_ms) = match delivery_mode {
+        BinaryDeliveryMode::HttpUpload {
+            endpoint,
+            timeout_secs,
+            max_retries,
+            retry_delay_ms,
+        } => (endpoint, *timeout_secs, *max_retries, *retry_delay_ms),
+        _ => {
+            return Err(RustCompilationError::SystemError(
+                "Invalid delivery mode for HTTP upload".to_string(),
+            ))
+        }
+    };
+
+    loop {
+        attempts += 1;
+
+        match upload_binary_to_endpoint(
+            common_config,
+            endpoint,
+            timeout_secs,
+            binary_path,
+            metadata,
+        )
+        .await
+        {
+            Ok(result) => {
+                if attempts > 1 {
+                    info!(
+                        "Binary upload succeeded on attempt {} for pipeline {} (program version: {})",
+                        attempts, metadata.pipeline_id, metadata.program_version
+                    );
+                }
+                return Ok(result);
+            }
+            Err(e) => {
+                if attempts > max_retries {
+                    error!(
+                        "Binary upload failed after {} attempts for pipeline {} (program version: {}): {:?}",
+                        attempts, metadata.pipeline_id, metadata.program_version, e
+                    );
+                    return Err(e);
+                }
+
+                warn!(
+                    "Binary upload attempt {} failed for pipeline {} (program version: {}): {:?}. Retrying in {}ms...",
+                    attempts, metadata.pipeline_id, metadata.program_version, e, retry_delay_ms
+                );
+
+                // Wait before retrying with exponential backoff
+                sleep(Duration::from_millis(retry_delay_ms)).await;
+                retry_delay_ms *= 2; // Double the delay for next attempt
+            }
+        }
+    }
+}
+
+/// Uploads the compiled binary to an HTTP endpoint (single attempt).
+async fn upload_binary_to_endpoint(
+    common_config: &CommonConfig,
+    endpoint: &str,
+    timeout_secs: u64,
+    binary_path: &Path,
+    metadata: &BinaryMetadata,
+) -> Result<String, RustCompilationError> {
+    // Read binary file
+    let binary_data = fs::read(binary_path).await.map_err(|e| {
```

Comment:
e.g., from earlier comments, this should be streaming or chunked so you don't have it all at once